systems/controllers/QPCommon.cpp

```diff
@@ -0,0 +1,741 @@
+
+      pdata->fqp = tmp.transpose()*Qy*D_ls*Jcom;
+      // mexPrintf("fqp head: %f %f %f\n", pdata->fqp(0), pdata->fqp(1), pdata->fqp(2));
+      pdata->fqp += tmp1.transpose()*tmp2;
+      pdata->fqp += (S*x_bar + 0.5*s1).transpose()*B_ls*Jcom;
+      pdata->fqp -= u0.transpose()*tmp2;
+      pdata->fqp -= y0.transpose()*Qy*D_ls*Jcom;
+      pdata->fqp -= (params->whole_body.w_qdd.array()*pid_out.qddot_des.array()).matrix().transpose();
+      if (include_angular_momentum) {
+        pdata->fqp += qd.transpose()*pdata->Akdot.transpose()*params->W_kdot*pdata->Ak;
+        pdata->fqp -= kdot_des.transpose()*params->W_kdot*pdata->Ak;
+      }
+      f.head(nq) = pdata->fqp.transpose();
+     } else {
+      f.head(nq) = -pid_out.qddot_des;
+    } 
+  }
+  f.tail(nf+neps) = VectorXd::Zero(nf+neps);
+  
+  int neq = 6+neps+6*n_body_accel_eq_constraints+qp_input->whole_body_data.num_constrained_dofs;
+  MatrixXd Aeq = MatrixXd::Zero(neq,nparams);
+  VectorXd beq = VectorXd::Zero(neq);
+  
+  // constrained floating base dynamics
+  //  H_float*qdd - J_float'*lambda - Dbar_float*beta = -C_float
+  Aeq.topLeftCorner(6,nq) = pdata->H_float;
+  beq.topRows(6) = -pdata->C_float;
+    
+  if (nc>0) {
+    Aeq.block(0,nq,6,nc*nd) = -D_float;
+  }
+  
+  if (nc > 0) {
+    // relative acceleration constraint
+    Aeq.block(6,0,neps,nq) = Jp;
+    Aeq.block(6,nq,neps,nf) = MatrixXd::Zero(neps,nf);  // note: obvious sparsity here
+    Aeq.block(6,nq+nf,neps,neps) = MatrixXd::Identity(neps,neps);             // note: obvious sparsity here
+    beq.segment(6,neps) = (-Jpdot -params->Kp_accel*Jp)*qd; 
+  }    
+  
+  // add in body spatial equality constraints
+  // VectorXd body_vdot;
+  MatrixXd orig = MatrixXd::Zero(4,1);
+  orig(3,0) = 1;
+  int equality_ind = 6+neps;
+  MatrixXd Jb(6,nq);
+  MatrixXd Jbdot(6,nq);
+  for (int i=0; i<desired_body_accelerations.size(); i++) {
+    if (desired_body_accelerations[i].weight < 0) { // negative implies constraint
+      if (!inSupport(active_supports,desired_body_accelerations[i].body_id0)) {
+        pdata->r->forwardJac(desired_body_accelerations[i].body_id0,orig,1,Jb);
+        pdata->r->forwardJacDot(desired_body_accelerations[i].body_id0,orig,1,Jbdot);
+
+        for (int j=0; j<6; j++) {
+          if (!std::isnan(desired_body_accelerations[i].body_vdot(j))) {
+            Aeq.block(equality_ind,0,1,nq) = Jb.row(j);
+            beq[equality_ind++] = -Jbdot.row(j)*qd + desired_body_accelerations[i].body_vdot(j);
+          }
+        }
+      }
+    }
+  }
+
+  if (qp_input->whole_body_data.num_constrained_dofs>0) {
+    // add joint acceleration constraints
+    for (int i=0; i<qp_input->whole_body_data.num_constrained_dofs; i++) {
+      Aeq(equality_ind,(int)condof[i]-1) = 1;
+      beq[equality_ind++] = pid_out.qddot_des[(int)condof[i]-1];
+    }
+  }  
+  
+  int n_ineq = 2*nu+2*6*desired_body_accelerations.size();
+  MatrixXd Ain = MatrixXd::Zero(n_ineq,nparams);  // note: obvious sparsity here
```

Comment:
agreed. But I think I might avoid doing that for this pull request, since that code is all directly copied from the existing controller. I'll make a new issue to do the sparse conversion. #836